File: 백준/Silver/15828. Router/Router.py
from collections import deque

N = int(input())
buffer = deque(maxlen=N)

while True:
    packet = int(input())
    if packet == -1:
        break

    if packet == 0 and buffer:              # packet이 0이고(처리) buffer가 비어 있지 않을 때
        buffer.popleft()
    elif packet != 0 and len(buffer) < N:   # buffer가 꽉 차지 않았을 때
        buffer.append(packet)

if buffer:                              # buffer가 비어 있지 않으면
    print(' '.join(map(str, buffer)))
else:                                   # buffer가 비어 있으면
    print('empty')

File: 백준/Silver/15828. Router/Router.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        Deque<Integer> buffer = new ArrayDeque<>();

        while(true) {
            int packet = Integer.parseInt(br.readLine());
            if(packet == -1) {
                break;
            }
            if(packet == 0) {
                buffer.removeFirst();
            }else if(buffer.size() < N) {
                buffer.addLast(packet);
            }
        }

        StringBuilder sb = new StringBuilder();

        if(buffer.isEmpty()) {
            System.out.println("empty");
        }else {
            int bufferSize = buffer.size();
            for(int i = 0; i < bufferSize; i++) {
                sb.append(buffer.removeFirst()).append(" ");
            }
            System.out.println(sb.toString());
        }


    }
}
